### PR TITLE
[PR] Implementing basic file upload

### LINKED ISSUE #51: Epic: Upload images to `AWS S3` via `Phoenix LiveView`
At present this repo is just a loose collection of notes and some sample code.
What it should be instead is an end-to-end working example of uploading, compressing, resizing, rotating and cropping images.

# Goal

Our goal with image uploads is to be able to seamlessly upload any type of image
have it saved in a reliable place e.g. `S3` (though we can research cheaper comparable options later...)
and display that image in our `App` and other places. 

We want to be able to upload in a `Web` interface and in a `Flutter` App.

# Todo

+ [x] Move all existing code to an `_archive` directory in case we need to refer to any of it.
  + [x] Move the `README.md` to `_archive` too so you have a fresh start.
  > **Note**: it will all be **`DELETED`** once we have fully working version. 
+ [x] Watch Chris McCord's `LiveView` Uploads Demo:

Phoenix LiveView Uploads Deep Dive: https://youtu.be/PffpT2eslH8
image

+ [x] Create a new `Phoenix` Project and replicate what Chris but document the steps 
  e.g: https://github.com/dwyl/phoenix-chat-example
+ [x] Upload images to `AWS S3`: https://gist.github.com/chrismccord/37862f1f8b1f5148644b75d20d1cb073
+ [x] Use a `Content ID` [`cid`](https://github.com/dwyl/cid) as the image name so that its unique.
+ [x] Once images are uploaded to `S3`, they should be "processed" by a `Lambda` function. 
  > **Note**: resizing and compressing images is a well-known/resolved challenge: 
  https://aws.amazon.com/blogs/compute/resize-images-on-the-fly-with-amazon-s3-aws-lambda-and-amazon-api-gateway/
+ [ ] In-browser resizing. This will be it's own `Epic` which will start with research for existing solutions. 🔍 

+ [x] Deploy the app to Fly.io: https://fly.io/dashboard/dwyl-image-uploads/ e.g: https://image-upload.fly.dev/ 🚀 

If we can get to the point where we are uploading images in a `Phoenix LiveView` App we will already be showing great progress.
Don't worry too much about the UI/UX at this stage. Use `TailwindUI`: 
https://tailwindui.com/components/application-ui/forms/form-layouts
image

@LuchoTurtle as discussed on standup this morning, this should be the next "big chunk" of work you should attempt to tackle.
If you have questions, please comment. 🙏

### PR DESCRIPTION
**Pull request by LuchoTurtle**
fixes #51

This PR will create a `LiveView` file upload page that will upload files to a `S3` bucket.

### CODE REVIEW

**Comment by LuchoTurtle** [comment]
@nelsonic This should now be mergeable.
The app is updated to https://imgup.fly.dev/. 

I've set lifecycle rules on both `S3` buckets so **all the contents are erased every day** to prevent abuse from people (because the `S3` buckets are publicly accessible to read - not to write, though). 
I've followed https://repost.aws/knowledge-center/s3-empty-bucket-lifecycle-rule for this.

I've already set the secrets needed for the deploy to work for the Github Actions, as well.

Everything should be documented in the `README`, including how it was deployed in `AWS`. 

This does *not* have in-browser resizing, as that'll be left for another issue (and as stated in https://github.com/dwyl/imgup/issues/51, it's an epic on its own).

**Comment by nelsonic** [comment]
```sh
mix c
==> ex_multihash
Compiling 2 files (.ex)
warning: defp is_valid_hash_code/1 is private, @doc attribute is always discarded for private functions/macros/types
  lib/multihash.ex:178: Multihash.is_valid_hash_code/1

warning: defp encode_internal/3 is private, @doc attribute is always discarded for private functions/macros/types
  lib/multihash.ex:185: Multihash.encode_internal/3

warning: defp decode_internal/3 is private, @doc attribute is always discarded for private functions/macros/types
  lib/multihash.ex:196: Multihash.decode_internal/3

warning: defp check_length/2 is private, @doc attribute is always discarded for private functions/macros/types
  lib/multihash.ex:209: Multihash.check_length/2

warning: defp check_truncated_digest_length/3 is private, @doc attribute is always discarded for private functions/macros/types
  lib/multihash.ex:219: Multihash.check_truncated_digest_length/3

warning: defp check_digest_length/2 is private, @doc attribute is always discarded for private functions/macros/types
  lib/multihash.ex:229: Multihash.check_digest_length/2

warning: defp get_hash_info/1 is private, @doc attribute is always discarded for private functions/macros/types
  lib/multihash.ex:239: Multihash.get_hash_info/1

warning: defp get_hash_function/1 is private, @doc attribute is always discarded for private functions/macros/types
  lib/multihash.ex:245: Multihash.get_hash_function/1

warning: defp get_from_dict/3 is private, @doc attribute is always discarded for private functions/macros/types
  lib/multihash.ex:251: Multihash.get_from_dict/3

warning: :crypto.hash/2 defined in application :crypto is used by the current application but the current application does not depend on :crypto. To fix this, you must do one of:

  1. If :crypto is part of Erlang/Elixir, you must include it under :extra_applications inside "def application" in your mix.exs

  2. If :crypto is a dependency, make sure it is listed under "def deps" in your mix.exs

  3. In case you don't want to add a requirement to :crypto, you may optionally skip this warning by adding [xref: [exclude: [:crypto]]] to your "def project" in mix.exs

Invalid call found at 3 locations:
  lib/util.ex:13: Multihash.Util.sum/2
  lib/util.ex:14: Multihash.Util.sum/2
  lib/util.ex:15: Multihash.Util.sum/2

warning: incompatible types:

    float() !~ integer()

in expression:

    # lib/multihash.ex:87
    <<hash_code>>

where "hash_code" was given the type integer() | float() in:

    # lib/multihash.ex:86
    is_number(hash_code)

where "hash_code" was given the type integer() in:

    # lib/multihash.ex:87
    <<hash_code>>

HINT: all expressions given to binaries are assumed to be of type integer() unless said otherwise. For example, <<expr>> assumes "expr" is an integer. Pass a modifier, such as <<expr::float>> or <<expr::binary>>, to change the default behaviour.

Conflict found at
  lib/multihash.ex:87: Multihash.encode/3

Generated ex_multihash app
==> b58
Compiling 1 file (.ex)
Generated b58 app
==> phoenix
Compiling 71 files (.ex)
Generated phoenix app
==> phoenix_live_view
Compiling 35 files (.ex)
Generated phoenix_live_view app
==> excid
Compiling 1 file (.ex)
Generated excid app
==> app
Compiling 7 files (.ex)
Generated app app
..23:55:57.865 [error] GenServer #PID<0.1745.0> terminating
** (System.EnvError) could not fetch environment variable "AWS_ACCESS_KEY_ID" because it is not set
    (elixir 1.14.3) lib/system.ex:706: System.fetch_env!/1
    (app 1.0.0) lib/app_web/live/imgup_live.ex:22: AppWeb.ImgupLive.presign_upload/2
    (phoenix_live_view 0.19.0) lib/phoenix_live_view/upload.ex:387: anonymous fn/3 in Phoenix.LiveView.Upload.external_preflight/4
    (elixir 1.14.3) lib/enum.ex:4751: Enumerable.List.reduce/3
    (elixir 1.14.3) lib/enum.ex:2514: Enum.reduce_while/3
    (phoenix_live_view 0.19.0) lib/phoenix_live_view/upload.ex:386: Phoenix.LiveView.Upload.external_preflight/4
    (phoenix_live_view 0.19.0) lib/phoenix_live_view/channel.ex:187: anonymous fn/7 in Phoenix.LiveView.Channel.handle_info/2
    (phoenix_live_view 0.19.0) lib/phoenix_live_view/channel.ex:1259: Phoenix.LiveView.Channel.write_socket/4
    (phoenix_live_view 0.19.0) lib/phoenix_live_view/channel.ex:180: Phoenix.LiveView.Channel.handle_info/2
    (stdlib 4.2) gen_server.erl:1123: :gen_server.try_dispatch/4
    (stdlib 4.2) gen_server.erl:1200: :gen_server.handle_msg/6
    (stdlib 4.2) proc_lib.erl:240: :proc_lib.init_p_do_apply/3
Last message: %Phoenix.Socket.Message{topic: "lv:phx-F2WWuT02t6vXWReC", event: "allow_upload", payload: %{"entries" => [%{"content" => <<137, 80, 78, 71, 13, 10, 26, 10, 0, 0, 0, 13, 73, 72, 68, 82, 0, 0, 2, 216, 0, 0, 0, 99, 8, 6, 0, 0, 0, 161, 65, 119, 154, 0, 0, 0, 25, 116, 69, 88, 116, 83, 111, 102, ...>>, "name" => "phoenix.png", "ref" => "6115", "relative_path" => nil, "size" => 13900, "type" => "image/png"}], "ref" => "phx-F2WWuT1BpTKcfRfC"}, ref: "1", join_ref: 0}
23:55:57.868 [error] GenServer #PID<0.1743.0> terminating
** (System.EnvError) could not fetch environment variable "AWS_ACCESS_KEY_ID" because it is not set
    (elixir 1.14.3) lib/system.ex:706: System.fetch_env!/1
    (app 1.0.0) lib/app_web/live/imgup_live.ex:22: AppWeb.ImgupLive.presign_upload/2
    (phoenix_live_view 0.19.0) lib/phoenix_live_view/upload.ex:387: anonymous fn/3 in Phoenix.LiveView.Upload.external_preflight/4
    (elixir 1.14.3) lib/enum.ex:4751: Enumerable.List.reduce/3
    (elixir 1.14.3) lib/enum.ex:2514: Enum.reduce_while/3
    (phoenix_live_view 0.19.0) lib/phoenix_live_view/upload.ex:386: Phoenix.LiveView.Upload.external_preflight/4
    (phoenix_live_view 0.19.0) lib/phoenix_live_view/channel.ex:187: anonymous fn/7 in Phoenix.LiveView.Channel.handle_info/2
    (phoenix_live_view 0.19.0) lib/phoenix_live_view/channel.ex:1259: Phoenix.LiveView.Channel.write_socket/4
    (phoenix_live_view 0.19.0) lib/phoenix_live_view/channel.ex:180: Phoenix.LiveView.Channel.handle_info/2
    (stdlib 4.2) gen_server.erl:1123: :gen_server.try_dispatch/4
    (stdlib 4.2) gen_server.erl:1200: :gen_server.handle_msg/6
    (stdlib 4.2) proc_lib.erl:240: :proc_lib.init_p_do_apply/3
Last message: {:EXIT, #PID<0.1740.0>, {%System.EnvError{env: "AWS_ACCESS_KEY_ID"}, [{System, :fetch_env!, 1, [file: 'lib/system.ex', line: 706]}, {AppWeb.ImgupLive, :presign_upload, 2, [file: 'lib/app_web/live/imgup_live.ex', line: 22]}, {Phoenix.LiveView.Upload, :"-external_preflight/4-fun-0-", 3, [file: 'lib/phoenix_live_view/upload.ex', line: 387]}, {Enumerable.List, :reduce, 3, [file: 'lib/enum.ex', line: 4751]}, {Enum, :reduce_while, 3, [file: 'lib/enum.ex', line: 2514]}, {Phoenix.LiveView.Upload, :external_preflight, 4, [file: 'lib/phoenix_live_view/upload.ex', line: 386]}, {Phoenix.LiveView.Channel, :"-handle_info/2-fun-2-", 7, [file: 'lib/phoenix_live_view/channel.ex', line: 187]}, {Phoenix.LiveView.Channel, :write_socket, 4, [file: 'lib/phoenix_live_view/channel.ex', line: 1259]}, {Phoenix.LiveView.Channel, :handle_info, 2, [file: 'lib/phoenix_live_view/channel.ex', line: 180]}, {:gen_server, :try_dispatch, 4, [file: 'gen_server.erl', line: 1123]}, {:gen_server, :handle_msg, 6, [file: 'gen_server.erl', line: 1200]}, {:proc_lib, :init_p_do_apply, 3, [file: 'proc_lib.erl', line: 240]}]}}


  1) test uploading a file (AppWeb.ImgupLiveTest)
     test/app_web/live/imgup_live_test.exs:76
     ** (EXIT from #PID<0.1740.0>) an exception was raised:
         ** (System.EnvError) could not fetch environment variable "AWS_ACCESS_KEY_ID" because it is not set
             (elixir 1.14.3) lib/system.ex:706: System.fetch_env!/1
             (app 1.0.0) lib/app_web/live/imgup_live.ex:22: AppWeb.ImgupLive.presign_upload/2
             (phoenix_live_view 0.19.0) lib/phoenix_live_view/upload.ex:387: anonymous fn/3 in Phoenix.LiveView.Upload.external_preflight/4
             (elixir 1.14.3) lib/enum.ex:4751: Enumerable.List.reduce/3
             (elixir 1.14.3) lib/enum.ex:2514: Enum.reduce_while/3
             (phoenix_live_view 0.19.0) lib/phoenix_live_view/upload.ex:386: Phoenix.LiveView.Upload.external_preflight/4
             (phoenix_live_view 0.19.0) lib/phoenix_live_view/channel.ex:187: anonymous fn/7 in Phoenix.LiveView.Channel.handle_info/2
             (phoenix_live_view 0.19.0) lib/phoenix_live_view/channel.ex:1259: Phoenix.LiveView.Channel.write_socket/4
             (phoenix_live_view 0.19.0) lib/phoenix_live_view/channel.ex:180: Phoenix.LiveView.Channel.handle_info/2
             (stdlib 4.2) gen_server.erl:1123: :gen_server.try_dispatch/4
             (stdlib 4.2) gen_server.erl:1200: :gen_server.handle_msg/6
             (stdlib 4.2) proc_lib.erl:240: :proc_lib.init_p_do_apply/3

..
Finished in 0.1 seconds (0.00s async, 0.1s sync)
5 tests, 1 failure

Randomized with seed 754153
----------------
COV    FILE                                        LINES RELEVANT   MISSED
100.0% lib/app.ex                                      9        0        0
100.0% lib/app/repo.ex                                 5        0        0
100.0% lib/app_web/components/layouts.ex               5        0        0
100.0% lib/app_web/endpoint.ex                        47        0        0
 46.2% lib/app_web/live/imgup_live.ex                 85       26       14
100.0% lib/app_web/router.ex                          27        2        0
  0.0% lib/app_web/s3_upload.ex                      127       24       24
[TOTAL]  26.9%
----------------
Generating report...
Saved to: cover/
FAILED: Expected minimum coverage of 100%, got 26.9%.
```

**Comment by nelsonic** [comment]
Gonna need to create some AWS Keys with the appropriate permissions tomorrow morning. 🧑‍💻

**Comment by nelsonic** [comment]
@SimonLab you have access to AWS (same role as @LuchoTurtle) so you should be able to run this on your `localhost`. 
Might be a good one to take a look at for `LiveView` refresh. 💭

**Comment by nelsonic** [comment]
Working on localhost after exporting AWS keys: 

image

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -55,6 +55,8 @@ jobs:
     if: github.ref == 'refs/heads/main'
     env:
       FLY_API_TOKEN: ${{ secrets.FLY_API_TOKEN }}
+      AWS_ACCESS_KEY_ID: ${{ secrets.AWS_ACCESS_KEY_ID }}
```
Comment (nelsonic) [inline .github/workflows/ci.yml:58]: @LuchoTurtle did you use your _own_ (`lucho-localhost`) AWS Access keys on Fly.io? 💭

Comment (LuchoTurtle) [reply .github/workflows/ci.yml:58]: Yes, I did. It was just to get the deployed app working to check if everything was working correctly and so it was easier on you to change these values in the repo secrets.

Comment (nelsonic) [reply .github/workflows/ci.yml:58]: Ok. Good to make progress but ideally we should just open an issue and assign these kinds of `DevOps` chores to me. 😜

File: lib/app_web/live/imgup_live.ex [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+     |> allow_upload(:image_list, accept: ~w(image/*), max_entries: 6, chunk_size: 64_000, max_file_size: 5_000_000, external: &presign_upload/2)}
+  end
+
+  # Adding presign for each entry for S3 upload --------
+
+  defp presign_upload(entry, socket) do
+    uploads = socket.assigns.uploads
+    bucket_original = "imgup-original"
+    bucket_compressed = "imgup-compressed"
+    key = Cid.cid("#{DateTime.utc_now() |> DateTime.to_iso8601()}_#{entry.client_name}")
```
Comment (nelsonic) [inline lib/app_web/live/imgup_live.ex:18]: the `key` should be the `cid` (hash) of the actual content, not a `DateTime`. 
please see: https://github.com/dwyl/imgup/issues/62

Using a `DateTime` means the filename will always be unique. We don't want that. 
We want to _avoid_ duplicates of files being uploaded.

**Pull request merged**